### DISCUSSION TITLE
Not able to convert DER public key Bio to EVP_Key

### DISCUSSION THREAD

**Discussion by chaitu-venky** [openssl/openssl#25438] [Q&A]
In my iOS application am trying to convert the public key data to EVP_KEY am using API "d2i_PUBKEY_bio" which is returning nil.

The error say's "not enough data". 

BIO *publicKeyBIO = <Public keydata> of 270 bytes size.
When i print data written to bio BIO_number_written(publicKeyBIO); prints 270 bytes.

EVP_PKEY *read_pkey = d2i_PUBKEY_bio(publicKeyBIO, NULL);

The result is read_pkey is NULL; Not sure what am missing here. 
It will be great if someone can help in this regard. Am using openssl 3.2.1

**Comment by mattcaswell** [comment]
This suggests that the data available is only partial in some way, i.e. the DER encoding has indicated that the expected length of the object is more than is actually available. This could be because the DER is corrupted in some way, or has somehow become truncated. It would be interesting to save those 270 bytes off to a file and then use the `asn1parse` app on it to confirm the data looks sane:

````
    openssl asn1parse -in myfile.der -inform DER
````

**Reply by chaitu-venky** [reply]
Thanks @mattcaswell for the quick response, I tried as you mentioned. 

Below is the response that it printed. 

```
    0:d=0  hl=4 l= 266 cons: SEQUENCE          
    4:d=1  hl=4 l= 257 prim: INTEGER           :C48CB56B942A53BE5DE6511510C60324F8F17BD003EF73E7D5559D1F53F9B2DEE3853BBD3D4AE885C67090DF9670164AADD6381CAFE5B59315F504F1F68BE545062F068933711EEA0859D3E4C93E7869F7DEAC09CF26047E286380A428CB4377464A07DE1ADEA11BE8CAE897D8829F780B826093E50C30F33D7F2D8E309AC11DCC11ADC59609558922D1BB85B731C201C2EAAF91BB890631BAAC5CC237B36193570C2D7FA30821E6C1C2E17399742454FC93F6E94315A57FBF8F21EBC78CCBC8FCEC78C9D716D60F69BC920724DC5CA9F8976F64A1F87C91CC46FA103F50FA29A120480878B3913AAFF09134EA16120253D68E9B454503AADB8DDC1A696A8BE5
  265:d=1  hl=2 l=   3 prim: INTEGER           :010001
```

[myfile.der.zip](https://github.com/user-attachments/files/16988479/myfile.der.zip)

**Reply by levitte** [reply]
That is a correct [RSAPublicKey](https://www.rfc-editor.org/rfc/rfc8017#appendix-A.1.1) structure, indeed.

Trouble is, `d2i_PUBKEY_bio` expects the more general SubjectPublicKeyInfo structure [^1].  To read what you have, you need to use `d2i_RSAPublicKey_bio` instead.

[^1]: SubjectPublicKeyInfo is defined in [RFC 3280](https://www.rfc-editor.org/rfc/rfc3280#appendix-A.1)

**Reply by chaitu-venky** [reply]
Currently am using openssl 3.2.1 d2i_RSAPublicKey_bio API is deprecated. So any other way that i can make this work?
